proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,256 @@
+# Dark Mode Extension Icons
+
+**Summary**
+
+Feature to enable developers to enhance extension icon visibility in dark mode.
+
+**Document Metadata**
+
+**Author:** &lt;solomonkinard>
+
+**Sponsoring Browser:** Chromium.
+
+**Contributors:** &lt;oliverdunk>
+
+**Created:** 2024-04-05
+
+**Related Issues:**
+* https://crbug.com/893175
+* https://github.com/w3c/webextensions/issues/229
+
+## Motivation
+
+### Objective
+
+Extension developers will be able to supply and define a set of icons to be used
+in the event that user has expressed the preference for a page that has a dark
+theme. The set of dark mode icons can and will be made automatically visible
+once the system setting is in dark mode, thereby enhancing the overall browser
+experience.
+
+#### Use Cases
+
+1. Improved icon visibility in the extension toolbar.
+1. Improved icon visibility on management the management and shortcuts pages.
+1. Dark mode icon declarations made possible through the extension manifest.
+1. setIcon() will allow setting of dark and/or light mode specific icons.
+1. Improved icon visibility on context menu (Chrome and Safari reply on default
+icons, only Firefox can specify icons).
+1. Improved icon visibility on side panel (Chrome replies on default icons,
+Firefox can specify icons).
+
+### Known Consumers
+
+The associated and linked bug has 48 stars and 31 comments.
+
+## Specification
+
+### Schema
+
+Some browsers don't currently have plans to support svg. Therefore, a drop-in
+replacement of an already supported image type will be allowed.
+
+Shared IDL
+```
+// DOMString <mode> : "dark" | "light";
+
+dictionary IconSizeToPath {
+  DOMString <size> => DOMString <path>;
+}
+
+dictionary IconVariants {
+  DOMString <mode> => IconSizeToPath;
+}
+
+dictionary IconVariantsKey {
```

Comment:
Unfortunately, icons aren't currently implemented in IDL. I decided to switch to IDL for this proposal to avoid inventing a new dialect. That said, can you provide a pointer -- my intent was to express what it it's supposed to be as DOMString isn't clear. It would be cleaner if there was a notion of `Path = DOMString` above, to then replace "DOMString <default_path>" here. The word default could be dropped as well. Any pointers on how to represent a DOMString with the aforementioned specificity in this proposal?

Reply:
In Firefox, we don't use IDL either, but JSON schemas for the manifest.

But if you want to be closer to IDL, one simplification aspect is to recognize that property names are always strings. So `DOMString <size>` could be simplified to `<size>` with a description of what it could be.

An alternative could be to use Typescript syntax, which is explicitly mentioned as a way to specify the API: https://github.com/w3c/webextensions/blob/main/proposals/proposal_template.md#schema

I don't mind the exact syntax as long as it is obvious and unambiguous.

Reply:
I found several examples of specifying manifest keys in idl in Chrome: https://cs.chromium.org/ManifestKeys%20file:extensions%20file:.idl

Reply:
tldr: TypeScript is used instead of IDL. ImageData is supported in action.setIcon() for icon_variants, and a future work section now includes exposing to developers whether the user preference is dark mode.

I decided to go for TypeScript instead. I also reinvented the inclusion of ImageData in setIcon for IconVariants. I think it's complex and yet easy to understand. However, implementation would also be complex. For that reason, I am not in favor of supporting ImageData in setIcon for anything related to icon_variants. However, it's included here in case no one else also objects to its inclusion, like me.

My argument against its inclusion is that setting ImageData is already a very advanced technique for developers and I think developers would be better serveed if they had an API endpoint to discover whether the user preference is dark mode or not. I'll add that new insight to the future work section.

Edit: On second thought, I could make a case not to expose the dark mode setting to developers. For that reason, I won't add it to the future work section at this time. That doesn't mean that I'm 100% behind ImageData support, it's just that I haven't thought of a better alternative so far. I don't like how complex ImageData is, how much work required to implement it, and no knowledge of how much it's even wanted by developers in the first place.

Reply:
<img width="478" alt="Screenshot 2024-04-11 at 12 24 20" src="https://github.com/w3c/webextensions/assets/776467/a0a86062-6f76-4dc7-bb38-ee8cdff9234a">

On macOS, Activity Monitor can show mini CPU usage graphic on its dock icon. Using imageData to dynamically set extension icon is similar.

I have two extensions that set icon with imageData dynamically. They use (offscreen) canvas to draw a graphic and set it to icon. SetIcon is triggered by events or timer.

```
function generateIcon() {
    context2D.clearRect(0, 0, 32, 32);
    // draw some thing
    return context2D.getImageData(0, 0, 32, 32);
}

let imageData = generateIcon();
browser.action.setIcon({ imageData: { 32: imageData }});

// when supporting icon variants, the code might become:
let darkIcon = generateIcon('dark');
let lightIcon = generateIcon('light');
browser.action.setIcon(...);
```

Supporting icon variants in `action.setIcon()` is tricky and complex, because parameters `path` and `imageData` are orthogonal to the icon variation. So I suggest the following API shape:

```
action.setIcon({
    path,
    pathVariants,
    imageData,
    imageDataVariants,
    tabId
});

// for example
action.setIcon({
    pathVariants: {
        light: { 32: 'light-32.png' },
        dark: { 32: 'dark-32.png' },
    }
});

action.setIcon({
    imageDataVariants: {
        light: { 32: lightImageData },
        dark: { 32: darkImageData },
    }
});
```

Reply:
The `flat_option` seems to be getting the most attention as of most recently.